### ISSUE TITLE
Sharing minion data cache causes false errors in returns

### ISSUE THREAD

**Issue by V3XATI0N** [saltstack/salt#43381]
### Description of Issue/Question
When multiple masters share a data cache (redis in my case), targeting a subset of minions with a module returns successfully for the targeted minions. However, all minions that were not included in the target also return with a `false` value, and this causes meta-processes watching salt's exit code for errors to think something has gone wrong.

Masters sharing a data cache should not behave this way. They should match the same behavior as if the master was using a local cache: appropriate returns for targeted minions, and nothing for minions that are not targeted.

### Steps to Reproduce Issue
1. Set up a salt master and a salt syndic reporting to that master.
2. Connect some minions to the master of masters, and some to the syndic.
3. On the master of masters, issue a command limited to a subset of minions (`salt -G grain:value test.ping`, for example).
4. Receive `True` returns for the targeted minions (where the `grain` grain is set to `value` here).
5. Also receive `False` returns for minions you did not target which should not be replying at all.
6. Weep.

### Versions Report
```
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: 1.10.0
       cherrypy: 10.2.1
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: 1.0.1
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.24.0
        libnacl: 1.4.5
       M2Crypto: Not Installed
           Mako: 1.0.3
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.24.0
         Python: 2.7.12 (default, Nov 19 2016, 06:48:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.2.0
           RAET: Not Installed
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: Ubuntu 16.04 xenial
         locale: UTF-8
        machine: x86_64
        release: 4.4.0-64-generic
         system: Linux
        version: Ubuntu 16.04 xenial
```

**Comment by V3XATI0N** [comment]
hello? is this thing on?

**Comment by gtmanfred** [comment]
@saltstack/team-core do we support having a shared minion cache between multi masters?

I seem to remember a discussion that the minion data cache was not built in a way that would take multiple writers to it in mind?

Thanks,
Daniel

**Comment by V3XATI0N** [comment]
According to [documentation](https://docs.saltstack.com/en/latest/topics/cache/index.html):
```
The pluggable data cache streamlines using various Salt topologies such as a Multi-Master or Salt Syndics configuration by allowing the data stored on the Salt Master about a Salt Minion to be available to other Salt Syndics or Salt Masters that a Salt Minion is connected to.
```
If masters aren't allowed to share the data cache, then what exactly is the point of sharing the data cache? If it's only intended for "multi-master" in active/active configuration, then what is the point of running Syndics? If there's no way to get Syndics to replicate cache data and have the master-of-masters behave as expected, then the entire Syndic system seems like a half-baked, useless feature set.

**Comment by gtmanfred** [comment]
@rallytime it looks like you wrote this portion of the documentation? Can you elaborate on it?

Thanks,
Daniel

**Comment by rallytime** [comment]
@gtmanfred Sure thing. I have been poking around to help with this issue and I _think_ what is going on is you're hitting a bug that was likely fixed by #43029. I'm curious if the problem you're seeing is fixed at the HEAD of the 2017.7.2 branch, or if there is a deeper issue as @gtmanfred was asking about above.

@DmitryKuzmenko I know you originally built the pluggable minion cache, so perhaps you have more insight into the original question. What do you think?

**Comment by DmitryKuzmenko** [comment]
Hmm... Interesting. I need to look at this. This shouldn't work in such way for the first look.
@V3XATI0N do you see the same behavior if
A) if you match minions by name, not grains
B) if you use the `localfs` cache driver
?

**Comment by V3XATI0N** [comment]
Thanks for jumping in. No, I don't see it if Salt uses `localfs`, in that case I only see output from the minions I'd expect -- except that if I target minions connected to a Syndic, I get nothing back because the cache isn't shared, so they return blank (hence my desire to use the shared cache).

Using shared cache, the `False` returns for non-targeted minions show up regardless of the method I use to target minions. Grain, glob, PCRE, Compound, etc. It makes no difference. Having a shared cache seems to trick the Master into outputting a response for everything, but only including data for targeted minions.

**Comment by DmitryKuzmenko** [comment]
Really strange. It's definitely a bug. Right now I have no time to reproduce it. @V3XATI0N could you please attach the non-sensitive parts of your config files to help us to reproduce and analyze the issue?

**Comment by V3XATI0N** [comment]
All of the masters in the environment share this master config template.

```
hash_type: sha256
log_level: error
log_level_logfile: debug

cache.redis.host: ## REDIS SERVER ##
cache.redis.port: 6505
cache.redis.db: '0'
cache.redis.password: ## REDIS AUTH PASSWORD ##
cache: redis

file_recv: True
file_recv_max_size: 500
file_roots:
  base:
    - /srv/salt
fileserver_backend:
  - git
  - roots
gitfs_provider: pygit2
gitfs_pillar_provider: pygit2
gitfs_pubkey: /etc/salt/pki/gitfs/salt-formula.pub
gitfs_privkey: /etc/salt/pki/gitfs/salt-formula
top_file_merging_strategy: same
gitfs_base: develop
gitfs_remotes:
  ## VARIOUS GITFS REMOTES ##

gitfs_ssl_verify: False
pillar_roots:
  base:
    - /srv/pillar
ext_pillar:
  ## GIT PILLAR THINGS ##
ext_pillar_first: True
pillar_gitfs_ssl_verify: False
pillar_source_merging_strategy: env_order
pillar_merge_lists: True
git_pillar_env: base
git_pillar_root: pillar
git_pillar_ssl_verify: False
git_pillar_pubkey: /etc/salt/pki/gitfs/salt-formula.pub
git_pillar_privkey: /etc/salt/pki/gitfs/salt-formula

{%- if grains['salt_role'] == "master-of-masters" %}
order_masters: True
{%- else %}
syndic_master: ## MASTER OF MASTERS ADDRESS ##
{%- endif %}

external_auth:
  pam:
    saltapi:
      - .*
      - '@runner'
      - '@wheel'

rest_cherrypy:
  port: 443
  ssl_crt: /etc/salt/pki/{{ grains['fqdn'] }}.crt
  ssl_key: /etc/salt/pki/{{ grains['fqdn'] }}.key
  webhook_url: /hook

reactor:
  - 'salt/cloud/*/created':
    - /srv/reactor/cloud-init.sls
```

**Comment by r0main** [comment]
Hi,

We hit the same issue on 2017.7.2. If my understanding is right, the PR #43670 has been merged and will correct this issue. Will the PR be included in 2017.7.3 ? If so, When the 2017.7.3 is supposed to be released publicly ?

Best regards.
